src/test/mir-opt/const_prop/aggregate/rustc.main.ConstProp.diff

```diff
@@ -14,44 +14,35 @@
           StorageLive(_1);                 // scope 0 at $DIR/aggregate.rs:5:9: 5:10
           StorageLive(_2);                 // scope 0 at $DIR/aggregate.rs:5:13: 5:24
           StorageLive(_3);                 // scope 0 at $DIR/aggregate.rs:5:13: 5:22
-          _3 = (const 0i32, const 1i32, const 2i32); // scope 0 at $DIR/aggregate.rs:5:13: 5:22
+          (_3.0: i32) = const 0i32;        // scope 0 at $DIR/aggregate.rs:5:13: 5:22
                                            // ty::Const
                                            // + ty: i32
                                            // + val: Value(Scalar(0x00000000))
                                            // mir::Constant
                                            // + span: $DIR/aggregate.rs:5:14: 5:15
                                            // + literal: Const { ty: i32, val: Value(Scalar(0x00000000)) }
+          (_3.1: i32) = const 1i32;        // scope 0 at $DIR/aggregate.rs:5:13: 5:22
                                            // ty::Const
                                            // + ty: i32
                                            // + val: Value(Scalar(0x00000001))
                                            // mir::Constant
                                            // + span: $DIR/aggregate.rs:5:17: 5:18
                                            // + literal: Const { ty: i32, val: Value(Scalar(0x00000001)) }
+          (_3.2: i32) = const 2i32;        // scope 0 at $DIR/aggregate.rs:5:13: 5:22
                                            // ty::Const
                                            // + ty: i32
                                            // + val: Value(Scalar(0x00000002))
                                            // mir::Constant
                                            // + span: $DIR/aggregate.rs:5:20: 5:21
                                            // + literal: Const { ty: i32, val: Value(Scalar(0x00000002)) }
--         _2 = (_3.1: i32);                // scope 0 at $DIR/aggregate.rs:5:13: 5:24
--         _1 = Add(move _2, const 0i32);   // scope 0 at $DIR/aggregate.rs:5:13: 5:28
-+         _2 = const 1i32;                 // scope 0 at $DIR/aggregate.rs:5:13: 5:24
```

Comment:
@oli-obk seems like even with your latest patch this still regresses?

Reply:
ok, this is getting weird. https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=387d515a8f94a2643d74169f9d80dfa5 is *basically* the same (except that there's an intial assignment to the entire local). Please try compiling this test with `RUSTC_LOG=rustc_mir::interpret::const_prop`, maybe that sheds some light on the problem

Reply:
I think you mean `rustc_mir::transform::const_prop`:
```
[TRACE rustc_mir::transform::const_prop] ConstProp starting for DefId(0:3 ~ aggregate[317d]::main[0])
[TRACE rustc_mir::transform::const_prop] local _1 will only be propagated inside its block, because it's a user variable
[TRACE rustc_mir::transform::const_prop] local _3 can't be propagated because of multiple assignments
[TRACE rustc_mir::transform::const_prop] local _3 can't be propagated because of multiple assignments
[TRACE rustc_mir::transform::const_prop] visit_statement: StorageLive(_1)
[TRACE rustc_mir::transform::const_prop] visit_statement: StorageLive(_2)
[TRACE rustc_mir::transform::const_prop] visit_statement: StorageLive(_3)
[TRACE rustc_mir::transform::const_prop] visit_statement: (_3.0: i32) = const 0i32
[TRACE rustc_mir::transform::const_prop] calling eval_rvalue_into_place(rvalue = const 0i32, place = (_3.0: i32))
[TRACE rustc_mir::transform::const_prop] can't propagate into (_3.0: i32)
[TRACE rustc_mir::transform::const_prop] visit_constant: const 0i32
[TRACE rustc_mir::transform::const_prop] visit_statement: (_3.1: i32) = const 1i32
[TRACE rustc_mir::transform::const_prop] calling eval_rvalue_into_place(rvalue = const 1i32, place = (_3.1: i32))
[TRACE rustc_mir::transform::const_prop] can't propagate into (_3.1: i32)
[TRACE rustc_mir::transform::const_prop] visit_constant: const 1i32
[TRACE rustc_mir::transform::const_prop] visit_statement: (_3.2: i32) = const 2i32
[TRACE rustc_mir::transform::const_prop] calling eval_rvalue_into_place(rvalue = const 2i32, place = (_3.2: i32))
[TRACE rustc_mir::transform::const_prop] can't propagate into (_3.2: i32)
[TRACE rustc_mir::transform::const_prop] visit_constant: const 2i32
[TRACE rustc_mir::transform::const_prop] visit_statement: _2 = (_3.1: i32)
[TRACE rustc_mir::transform::const_prop] calling eval_rvalue_into_place(rvalue = (_3.1: i32), place = _2)
[TRACE rustc_mir::transform::const_prop] propagation into _2 failed.
                            Nuking the entire site from orbit, it's the only way to be sure
[TRACE rustc_mir::transform::const_prop] visit_statement: _1 = Add(move _2, const 0i32)
[TRACE rustc_mir::transform::const_prop] checking BinaryOp(op = Add, left = move _2, right = const 0i32)
[TRACE rustc_mir::transform::const_prop] propagation into _1 failed.
                            Nuking the entire site from orbit, it's the only way to be sure
[TRACE rustc_mir::transform::const_prop] visit_constant: const 0i32
[TRACE rustc_mir::transform::const_prop] visit_statement: StorageDead(_2)
[TRACE rustc_mir::transform::const_prop] visit_statement: StorageDead(_3)
[TRACE rustc_mir::transform::const_prop] visit_statement: _0 = const ()
[TRACE rustc_mir::transform::const_prop] calling eval_rvalue_into_place(rvalue = const (), place = _0)
[TRACE rustc_mir::transform::const_prop] replacing const () with OpTy { op: Immediate(Scalar(<ZST>)), layout: TyAndLayout { ty: (), layout: Layout { fields: Arbitrary { offsets: [], memory_index: [] }, variants: Single { index: 0 }, abi: Aggregate { sized: true }, largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 0 }, pref: Align { pow2: 3 } }, size: Size { raw: 0 } } } }
[TRACE rustc_mir::transform::const_prop] skipping replace of Rvalue::Use(const () because it is already a const
[TRACE rustc_mir::transform::const_prop] propagated into _0
[TRACE rustc_mir::transform::const_prop] visit_constant: const ()
[TRACE rustc_mir::transform::const_prop] visit_statement: StorageDead(_1)
[TRACE rustc_mir::transform::const_prop] ConstProp done for DefId(0:3 ~ aggregate[317d]::main[0])
```

Reply:
ok thanks, this makes it clear what the problem is, I'll look into what the solution is.

Reply:
I can't do this locally without your PR. There's no way to produce the relevant MIR without your PR :(

I believe it should get fixed by changing https://github.com/rust-lang/rust/blob/4b1f86adbe41e8dd4864ca2315f43953dd503bb5/src/librustc_mir/transform/const_prop.rs#L790-L803 to change `ConstPropMode` to `OnlyInsideOwnBlock` if it currently is `FullConstProp`.

Reply:
Feel free to push to my branch -- we can make this a collaborative PR, maybe @jonas-schievink can review then.

Reply:
you need to check the box on the right allowing edits or give me access to your repo in order for me to push